### PR TITLE
Update policy for kafka gateway

### PR DESCRIPTION
**Pull request by phiz71**
**Issue**

https://gravitee.atlassian.net/browse/APIM-7143

**Description**

- use new HttpSecurityPolicy and BaseExecutionContext interface (BREAKING CHANGE: requires APIM 4.6+)
- implement kafka policy

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `4.0.0-APIM-7143-impl-security-chain-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/policy/gravitee-policy-oauth2/4.0.0-APIM-7143-impl-security-chain-SNAPSHOT/gravitee-policy-oauth2-4.0.0-APIM-7143-impl-security-chain-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java [hunk 1]
```diff
@@ -278,21 +333,33 @@ private Completable introspectAndValidateAccessToken(HttpExecutionContext ctx, S
      *      error="invalid_token",
      *      error_description="The access token expired"
      */
-    private Completable sendError(HttpExecutionContext ctx, String responseKey) {
-        String headerValue = BEARER_AUTHORIZATION_TYPE + " realm=\"gravitee.io\"";
+    private Completable sendError(BaseExecutionContext ctx, int errorStatus, String responseKey, String failureMessage) {
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:336]: We could take the opportunity to add an enum that could hold status, key and message.

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java [hunk 2]
```diff
@@ -278,21 +333,33 @@ private Completable introspectAndValidateAccessToken(HttpExecutionContext ctx, S
...
+            String headerValue = BEARER_AUTHORIZATION_TYPE + " realm=\"gravitee.io\"";
 
-        ctx.response().headers().add(HttpHeaderNames.WWW_AUTHENTICATE, headerValue);
+            httpPlainExecutionContext.response().headers().add(HttpHeaderNames.WWW_AUTHENTICATE, headerValue);
 
-        return ctx.interruptWith(new ExecutionFailure(UNAUTHORIZED_401).key(responseKey).message(OAUTH2_UNAUTHORIZED_MESSAGE));
+            return httpPlainExecutionContext.interruptWith(new ExecutionFailure(errorStatus).key(responseKey).message(failureMessage));
+        }
+        // FIXME: Kafka Gateway - manage interruption with Kafka.
+        return Completable.error(new Exception(responseKey));
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:345]: Can you create a ticket to handle the interruption with Kafka and point to these fixme so we don't miss them.

**Review by jhaeyaert — APPROVED** [review]
Nice!
I've approved and let you have a look at the 2 minor comments I left.

File: src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java (hunk 2 above)
Comment (phiz71) [reply src/main/java/io/gravitee/policy/oauth2/Oauth2Policy.java:345]: https://gravitee.atlassian.net/browse/APIM-7533

**Pull request merged**